File: src/Parser.ts
import { Token } from "./tokens";
import { TokenType, useTokenType } from "./tokens/TokenType";
import ExpressionNode from "./AST/ExpressionNode";
import StatementsNode from "./AST/StatementsNode";
import NumberNode from "./AST/NumberNode";
import VariableNode from "./AST/VariableNode";
import BinOperationNode from "./AST/BinOperationNode";
import UnarOperationNode from "./AST/UnarOperationNode";
import StringNode from "./AST/StringNode";
import IfNode from "./AST/IfNode";
import FunctionDeclarationNode from "./AST/FunctionDeclarationNode";
import FunctionCallNode from "./AST/FunctionCallNode";
import ReturnNode from "./AST/ReturnNode";

export default class Parser {
	tokens: Token[];
	pos: number = 0;
	source: string; // <-- add source to hold code text
	private getTokenType;
	private TokenNames;

	constructor(tokens: Token[], source: string) {
		const { getTokenType, TokenNames, tokenTypesList } = useTokenType();

		// <-- update constructor signature
		this.tokens = tokens;
		this.source = source;
		this.getTokenType = getTokenType;
		this.TokenNames = TokenNames;
	}

	// New helper to compute error details with column info
	private getErrorDetails(pos: number): string {
		// Adjust pos by subtracting one to correctly report the line
		const adjustedPos = pos > 0 ? pos - 1 : pos;
		const lines = this.source.split("\n");
		let currentPos = 0;
		for (let i = 0; i < lines.length; i++) {
			const line = lines[i];
			if (currentPos + line.length >= adjustedPos) {
				const column = adjustedPos - currentPos;
				return (
					`Line ${i + 1}, Column ${column}: ${line}\n` +
					`${" ".repeat(column)}^`
				);
			}
			currentPos += line.length + 1; // include newline
		}
		return "Unknown location";
	}

	// Method to match tokens
	match(...expected: TokenType[]): Token | null {
		if (this.pos < this.tokens.length) {
			const currentToken = this.tokens[this.pos];
			if (expected.find((type) => type.name === currentToken.type.name)) {
				this.pos += 1;
				return currentToken;
			}
		}
		return null;
	}

	// Updated require() that shows error details including the token that came before the failure.
	require(...expected: TokenType[]): Token {
		const token = this.match(...expected);
		if (!token) {
			const tokenPos = this.tokens[this.pos]?.pos || 0;
			const errorDetails = this.getErrorDetails(tokenPos);
			// If possible, show the text of the previous token
			const prevText = this.pos > 0 ? this.tokens[this.pos - 1].text : "";
			throw new Error(
				`Syntax Error at ${errorDetails} - Expected ${expected
					.map((t) => t.name)
					.join(" or ")} after '${prevText}'.`
			);
		}
		return token;
	}

	// Parse variable or data types
	parseVariableOrDataTypes(): ExpressionNode {
		const number = this.match(this.getTokenType(this.TokenNames.NUMBER));
		if (number != null) {
			return new NumberNode(number);
		}
		const variable = this.match(
			this.getTokenType(this.TokenNames.VARIABLE)
		);
		if (variable != null) {
			// Проверка на вызов функции
			if (this.match(this.getTokenType(this.TokenNames.LPAR)) != null) {
				this.pos -= 1;
				return this.parseFunctionCall(variable);
			}
			return new VariableNode(variable);
		}
		const string = this.match(this.getTokenType(this.TokenNames.STRING));
		if (string != null) {
			return new StringNode(string);
		}
		throw new Error(
			`Expecting a variable, number, or string at position: ${this.pos}`
		);
	}

	// Parse print statement
	parsePrint(): ExpressionNode {
		const operatorLog = this.match(this.getTokenType(this.TokenNames.LOG));
		if (operatorLog != null) {
			return new UnarOperationNode(operatorLog, this.parseFormula());
		}
		throw new Error(`Expecting a unary operator at position: ${this.pos}`);
	}

	// Parse parentheses
	parseParentheses(): ExpressionNode {
		if (this.match(this.getTokenType(this.TokenNames.LPAR)) != null) {
			const node = this.parseFormula();
			this.require(this.getTokenType(this.TokenNames.RPAR));

			return node;
		} else {
			return this.parseVariableOrDataTypes();
		}
	}

	// Parse braces
	parseBraces(): ExpressionNode {
		if (this.match(this.getTokenType(this.TokenNames.LBRACE)) != null) {
			const node = this.parseContext();

			this.pos -= 1;

			this.require(this.getTokenType(this.TokenNames.RBRACE));
			return node;
		} else {
			throw new Error(`Expecting { at position: ${this.pos}`);
		}
	}

	// New helper: parses parenthesized expressions or basic data types
	parsePrimary(): ExpressionNode {
		if (this.match(this.getTokenType(this.TokenNames.LPAR)) != null) {
			const node = this.parseFormula(); // later levels handle operations
			this.require(this.getTokenType(this.TokenNames.RPAR));
			return node;
		}
		return this.parseVariableOrDataTypes();
	}

	// New helper: handles multiplication and division operators
	parseMultiplicative(): ExpressionNode {
		let node = this.parsePrimary();
		let operator = this.match(
			this.getTokenType(this.TokenNames.MULT),
			this.getTokenType(this.TokenNames.DIV)
		);
		while (operator != null) {
			const rightNode = this.parsePrimary();
			node = new BinOperationNode(operator, node, rightNode);
			operator = this.match(
				this.getTokenType(this.TokenNames.MULT),
				this.getTokenType(this.TokenNames.DIV)
			);
		}
		return node;
	}

	// New helper: handles addition and subtraction operators
	parseAdditive(): ExpressionNode {
		let node = this.parseMultiplicative();
		let operator = this.match(
			this.getTokenType(this.TokenNames.PLUS),
			this.getTokenType(this.TokenNames.MINUS)
		);
		while (operator != null) {
			const rightNode = this.parseMultiplicative();
			node = new BinOperationNode(operator, node, rightNode);
			operator = this.match(
				this.getTokenType(this.TokenNames.PLUS),
				this.getTokenType(this.TokenNames.MINUS)
			);
		}
		return node;
	}

	// New helper: handles comparisons (=, <, >, etc.)
	parseComparison(): ExpressionNode {
		let node = this.parseAdditive();
		let operator = this.match(
			this.getTokenType(this.TokenNames.EQUAL),
			this.getTokenType(this.TokenNames.LESS),
			this.getTokenType(this.TokenNames.MORE),
			this.getTokenType(this.TokenNames.LESSEQ),
			this.getTokenType(this.TokenNames.MOREQ)
		);
		while (operator != null) {
			const rightNode = this.parseAdditive();
			node = new BinOperationNode(operator, node, rightNode);
			operator = this.match(
				this.getTokenType(this.TokenNames.EQUAL),
				this.getTokenType(this.TokenNames.LESS),
				this.getTokenType(this.TokenNames.MORE),
				this.getTokenType(this.TokenNames.LESSEQ),
				this.getTokenType(this.TokenNames.MOREQ)
			);
		}
		return node;
	}

	// Update parseFormula() to start with comparisons (the top arithmetic level)
	parseFormula(): ExpressionNode {
		return this.parseComparison();
	}

	// Parse function declaration
	parseFunctionDeclaration(): ExpressionNode {
		this.require(this.getTokenType(this.TokenNames.FUNCTION)); // Expecting 'HAM'
		const name = this.require(this.getTokenType(this.TokenNames.VARIABLE)); // Function name
		this.require(this.getTokenType(this.TokenNames.LPAR)); // Expecting '('

		const params: Token[] = [];
		if (this.match(this.getTokenType(this.TokenNames.RPAR)) == null) {
			do {
				params.push(
					this.require(this.getTokenType(this.TokenNames.VARIABLE))
				);
			} while (
				this.match(this.getTokenType(this.TokenNames.COMMA)) != null
			);
			this.require(this.getTokenType(this.TokenNames.RPAR)); // Expecting ')'
		}

		const body = this.parseBraces() as StatementsNode; // Function body

		return new FunctionDeclarationNode(name, params, body);
	}

	// Parse function call
	parseFunctionCall(name: Token): ExpressionNode {
		this.require(this.getTokenType(this.TokenNames.LPAR)); // Expecting '('

		const args: ExpressionNode[] = [];
		if (this.match(this.getTokenType(this.TokenNames.RPAR)) == null) {
			do {
				args.push(this.parseFormula());
			} while (
				this.match(this.getTokenType(this.TokenNames.COMMA)) != null
			);
			this.require(this.getTokenType(this.TokenNames.RPAR)); // Expecting ')'
		}

		return new FunctionCallNode(name, args);
	}

	// Parse return statement
	parseReturn(): ExpressionNode {
		// Already matched the 'RETURN' token in parseExpression.
		const token = this.require(this.getTokenType(this.TokenNames.RETURN));
		const expr = this.parseFormula();
		// Optionally require a semicolon if needed:
		this.require(this.getTokenType(this.TokenNames.SEMICOLON));
		return new ReturnNode(expr);
	}

	// Parse expressions
	parseExpression(): ExpressionNode {
		const currentToken = this.tokens[this.pos];
		if (
			currentToken.type.name ===
			this.getTokenType(this.TokenNames.FUNCTION).name
		) {
			return this.parseFunctionDeclaration();
		}
		if (
			currentToken.type.name ===
			this.getTokenType(this.TokenNames.RETURN).name
		) {
			return this.parseReturn();
		}

		// Other checks
		if (this.match(this.getTokenType(this.TokenNames.VARIABLE)) === null) {
			if (this.match(this.getTokenType(this.TokenNames.LOG)) !== null) {
				this.pos -= 1;
				return this.parsePrint();
			}
			if (this.match(this.getTokenType(this.TokenNames.IF))) {
				this.pos -= 1;
				return this.parseIfStatement();
			}
		}

		this.pos -= 1;
		let variableNode = this.parseVariableOrDataTypes();
		const assignOperator = this.match(
			this.getTokenType(this.TokenNames.ASSIGN)
		);
		if (assignOperator != null) {
			const rightFormulaNode = this.parseFormula();
			return new BinOperationNode(
				assignOperator,
				variableNode,
				rightFormulaNode
			);
		}

		return variableNode;
	}

	// Parse if-else statement
	parseIfStatement(): ExpressionNode {
		this.require(this.getTokenType(this.TokenNames.IF)); // Expecting 'NEU'
		const condition = this.parseFormula(); // Parsing the condition

		// Ensure that we expect a left curly brace
		if (this.match(this.getTokenType(this.TokenNames.LBRACE)) === null) {
			throw new Error(`Expected '{' at position: ${this.pos}`);
		}

		const trueBlock = this.parseContext(); // Parsing the true block

		// Move back one position because parseContext uses the match method to check for RBRACE
		this.pos -= 1;

		this.require(this.getTokenType(this.TokenNames.RBRACE)); // Expecting '}'

		let falseBlock: ExpressionNode | null = null; // Initializing the false block
		if (this.match(this.getTokenType(this.TokenNames.ELSE)) !== null) {
			// Checking for 'KO THI'
			if (this.match(this.getTokenType(this.TokenNames.IF)) !== null) {
				// If 'else if' exists
				this.pos -= 1; // Move back one position because we use the match method
				falseBlock = this.parseIfStatement(); // Recursively parse else if
			} else {
				if (
					this.match(this.getTokenType(this.TokenNames.LBRACE)) ===
					null
				) {
					throw new Error(`Expected '{' at position: ${this.pos}`);
				}
				falseBlock = this.parseContext(); // Parsing the false block

				// Move back one position because parseContext uses the match method to check for RBRACE
				this.pos -= 1;

				this.require(this.getTokenType(this.TokenNames.RBRACE)); // Expecting '}'
			}
		}

		return new IfNode(condition, trueBlock, falseBlock); // Creating IfNode
	}

	// Parse context
	parseContext(): ExpressionNode {
		const root = new StatementsNode();

		while (this.pos < this.tokens.length) {
			if (this.match(this.getTokenType(this.TokenNames.RBRACE)) !== null)
				break;

			const codeStringNode = this.parseExpression();

			// If the expression is an IF construct, function declaration, or return,
			// a semicolon is not required.
			if (
				codeStringNode instanceof IfNode ||
				codeStringNode instanceof FunctionDeclarationNode ||
				codeStringNode instanceof ReturnNode
			) {
				root.addNode(codeStringNode);
			} else {
				this.require(this.getTokenType(this.TokenNames.SEMICOLON)); // Otherwise, expect ';'
				root.addNode(codeStringNode);
			}
		}
		return root;
	}

	parseCode(): ExpressionNode {
		return this.parseContext();
	}
}

File: go/internal/parser/parser.go
package parser

import (
	"fmt"
	"strconv"

	"github.com/hoachnt/yanghoscript/internal/ast"
	"github.com/hoachnt/yanghoscript/internal/lexer"
)

type Parser struct {
	tokens []lexer.Token
	pos    int
}

func New(tokens []lexer.Token) *Parser {
	return &Parser{tokens: tokens, pos: 0}
}

func (p *Parser) match(types ...lexer.TokenType) *lexer.Token {
	if p.pos < len(p.tokens) {
		currentToken := p.tokens[p.pos]
		for _, t := range types {
			if currentToken.Type == t {
				p.pos++
				return &currentToken
			}
		}
	}
	return nil
}

func (p *Parser) require(types ...lexer.TokenType) *lexer.Token {
	token := p.match(types...)
	if token == nil {
		panic(fmt.Sprintf("Syntax Error: Expected %v at position %d", types, p.pos))
	}
	return token
}

func (p *Parser) parseVariableOrDataTypes() ast.ExpressionNode {
	if number := p.match(lexer.NUMBER); number != nil {
		val, _ := strconv.ParseFloat(number.Literal, 64)
		return &ast.NumberNode{Value: val}
	}

	if variable := p.match(lexer.IDENT); variable != nil {
		return &ast.VariableNode{Name: variable.Literal}
	}

	if str := p.match(lexer.STRING); str != nil {
		v := str.Literal
		if len(v) >= 2 && v[0] == '\'' && v[len(v)-1] == '\'' {
			v = v[1 : len(v)-1]
		}
		return &ast.StringNode{Value: v}
	}

	panic(fmt.Sprintf("Expected variable, number, or string at position %d, got token: %+v", p.pos, p.tokens[p.pos]))
}

func (p *Parser) parsePrint() ast.Node {
	logTok := p.require(lexer.LOG)
	return &ast.UnarOperationNode{
		Operator: string(logTok.Type),
		Operand:  p.parseFormula(),
	}
}

func (p *Parser) parseParamList() []string {
	params := []string{}
	for p.match(lexer.IDENT) != nil {
		params = append(params, p.tokens[p.pos-1].Literal)
		if p.match(lexer.COMMA) == nil {
			break
		}
	}
	return params
}

func (p *Parser) parseFunction() ast.Node {
	p.require(lexer.FUNCTION)
	if p.pos < len(p.tokens) && p.tokens[p.pos].Type == lexer.LPAREN {
		p.require(lexer.LPAREN)
		params := p.parseParamList()
		p.require(lexer.RPAREN)
		p.require(lexer.LBRACE)
		body := p.parseContext()
		p.require(lexer.RBRACE)
		return &ast.FunctionDeclarationNode{Name: "", Parameters: params, Body: body}
	}

	name := p.require(lexer.IDENT)
	p.require(lexer.LPAREN)
	params := p.parseParamList()
	p.require(lexer.RPAREN)
	p.require(lexer.LBRACE)
	body := p.parseContext()
	p.require(lexer.RBRACE)
	return &ast.FunctionDeclarationNode{Name: name.Literal, Parameters: params, Body: body}
}

func (p *Parser) parseBind() ast.Node {
	p.require(lexer.BIND)
	variable := p.require(lexer.IDENT)
	p.require(lexer.ASSIGN)
	value := p.parseFormula()
	return &ast.AssignNode{
		Variable:  &ast.VariableNode{Name: variable.Literal},
		Value:     value,
		Immutable: true,
	}
}

func (p *Parser) parseListLiteral() ast.ExpressionNode {
	p.require(lexer.LBRACKET)
	elements := []ast.ExpressionNode{}
	if p.match(lexer.RBRACKET) != nil {
		return &ast.ListNode{Elements: elements}
	}
	for {
		elements = append(elements, p.parseFormula())
		if p.match(lexer.RBRACKET) != nil {
			break
		}
		if p.match(lexer.COMMA) == nil {
			p.require(lexer.RBRACKET)
			break
		}
	}
	return &ast.ListNode{Elements: elements}
}

func (p *Parser) parsePrimary() ast.ExpressionNode {
	if p.match(lexer.LPAREN) != nil {
		node := p.parseFormula()
		p.require(lexer.RPAREN)
		return node
	}

	if p.pos < len(p.tokens) && p.tokens[p.pos].Type == lexer.LBRACKET {
		return p.parseListLiteral()
	}

	if p.pos < len(p.tokens) && p.tokens[p.pos].Type == lexer.FUNCTION {
		n := p.parseFunction()
		fd, ok := n.(*ast.FunctionDeclarationNode)
		if !ok || fd.Name != "" {
			panic("expected lambda expression (THE (args) ME ... MAY)")
		}
		return fd
	}

	if ident := p.match(lexer.IDENT); ident != nil {
		if p.pos < len(p.tokens) && p.tokens[p.pos].Type == lexer.LPAREN {
			return p.parseFunctionCall(ident).(ast.ExpressionNode)
		}
		return &ast.VariableNode{Name: ident.Literal}
	}

	return p.parseVariableOrDataTypes()
}

func (p *Parser) parseMultiplicative() ast.ExpressionNode {
	node := p.parsePrimary()
	for {
		if operator := p.match(lexer.MULTIPLY, lexer.DIVIDE); operator != nil {
			right := p.parsePrimary()
			node = &ast.BinOperationNode{Left: node, Operator: operator.Literal, Right: right}
			continue
		}
		break
	}
	return node
}

func (p *Parser) parseAdditive() ast.ExpressionNode {
	node := p.parseMultiplicative()
	for {
		if operator := p.match(lexer.PLUS, lexer.MINUS); operator != nil {
			right := p.parseMultiplicative()
			node = &ast.BinOperationNode{Left: node, Operator: operator.Literal, Right: right}
			continue
		}
		break
	}
	return node
}

func (p *Parser) parseIf() ast.ExpressionNode {
	p.require(lexer.IF)
	p.require(lexer.LPAREN)
	condition := p.parseFormula()
	p.require(lexer.RPAREN)
	if p.match(lexer.LBRACE) == nil {
		panic(fmt.Sprintf("Expected block at position: %d", p.pos))
	}
	trueBlock := p.parseContext()
	p.require(lexer.RBRACE)

	var falseBlock ast.ExpressionNode
	if p.match(lexer.ELSE) != nil {
		if p.match(lexer.IF) != nil {
			p.pos--
			falseBlock = p.parseIf()
		} else {
			if p.match(lexer.LBRACE) == nil {
				panic(fmt.Sprintf("Expected block at position: %d", p.pos))
			}
			falseBlock = p.parseContext()
			p.require(lexer.RBRACE)
		}
	}

	return &ast.IfNode{Condition: condition, ThenBranch: trueBlock, ElseBranch: falseBlock}
}

func (p *Parser) parseComparison() ast.ExpressionNode {
	node := p.parseAdditive()
	for {
		if operator := p.match(lexer.EQUAL, lexer.LESS, lexer.GREATER, lexer.LESSEQ, lexer.MOREQ, lexer.NOTEQUAL); operator != nil {
			right := p.parseAdditive()
			node = &ast.BinOperationNode{Left: node, Operator: operator.Literal, Right: right}
			continue
		}
		break
	}
	return node
}

func (p *Parser) parseFormula() ast.ExpressionNode {
	return p.parseComparison()
}

func (p *Parser) parseFunctionCall(name *lexer.Token) ast.Node {
	p.require(lexer.LPAREN)
	arguments := []ast.ExpressionNode{}
	if p.match(lexer.RPAREN) != nil {
		return &ast.FunctionCallNode{Name: name.Literal, Arguments: arguments}
	}
	for {
		arguments = append(arguments, p.parseFormula())
		if p.match(lexer.RPAREN) != nil {
			break
		}
		if p.match(lexer.COMMA) == nil {
			p.require(lexer.RPAREN)
			break
		}
	}
	return &ast.FunctionCallNode{Name: name.Literal, Arguments: arguments}
}

func (p *Parser) parseReturn() ast.Node {
	p.require(lexer.RETURN)
	value := p.parseFormula()
	return &ast.ReturnNode{Value: value}
}

func (p *Parser) parseContext() *ast.StatementsNode {
	statements := &ast.StatementsNode{}

	for p.pos < len(p.tokens) {
		if p.match(lexer.SEMICOLON) != nil {
			continue
		}

		if p.match(lexer.RBRACE) != nil {
			p.pos--
			break
		}

		if p.tokens[p.pos].Type == lexer.EOF {
			break
		}

		// Let parseIf consume THOI/KOTHI (else) for the surrounding NEU.
		if p.tokens[p.pos].Type == lexer.ELSE {
			break
		}

		switch p.tokens[p.pos].Type {
		case lexer.LOG:
			statements.Statements = append(statements.Statements, p.parsePrint())
		case lexer.IF:
			statements.Statements = append(statements.Statements, p.parseIf())
		case lexer.FUNCTION:
			statements.Statements = append(statements.Statements, p.parseFunction())
		case lexer.RETURN:
			statements.Statements = append(statements.Statements, p.parseReturn())
		case lexer.BIND:
			statements.Statements = append(statements.Statements, p.parseBind())
		case lexer.IDENT:
			name := p.tokens[p.pos]
			if p.pos+1 < len(p.tokens) && p.tokens[p.pos+1].Type == lexer.ASSIGN {
				panic("immutable bind: use CHOT name = value (không gán lại bằng '=' trần)")
			}
			if p.pos+1 < len(p.tokens) && p.tokens[p.pos+1].Type == lexer.LPAREN {
				p.pos++
				statements.Statements = append(statements.Statements, p.parseFunctionCall(&name))
			} else {
				panic(fmt.Sprintf("expected call or CHOT at position %d", p.pos))
			}
		default:
			panic(fmt.Sprintf("unexpected token in block: %v at %d", p.tokens[p.pos].Type, p.pos))
		}
	}

	return statements
}

func (p *Parser) ParseCode() (node *ast.StatementsNode, err error) {
	if len(p.tokens) == 0 {
		return nil, fmt.Errorf("no tokens to parse")
	}

	defer func() {
		if r := recover(); r != nil {
			switch v := r.(type) {
			case string:
				err = fmt.Errorf("%s", v)
			case error:
				err = v
			default:
				err = fmt.Errorf("%v", v)
			}
			node = nil
		}
	}()

	node = p.parseContext()
	return node, err
}
